paasta_tools/contrib/check_orphans.py

```diff
@@ -212,43 +213,54 @@ def get_instance_data(
-    for orphan in orphans:
-        orphans_by_host[orphan.host].append((orphan.port, orphan.service))
-
-    # collisions
-    instance_by_addr: DefaultDict[Tuple[str, int], Set[str]] = defaultdict(set)
-    for nerve_inst in nerve_instance_data:
-        instance_by_addr[(nerve_inst.host, nerve_inst.port)].add(nerve_inst.service)
-
-    collisions: List[str] = []
-    for zk_inst in zk_instance_data:
-        nerve_services = instance_by_addr[(zk_inst.host, zk_inst.port)]
-        if len(nerve_services) >= 1 and zk_inst.service not in nerve_services:
-            collisions.append(
-                f"[{zk_inst.host}:{zk_inst.port}] {zk_inst.service} collides with {nerve_services}"
-            )
 
-    if collisions:
-        logger.warning("Collisions found! Traffic is being misrouted!")
-        print("\n".join(collisions))
-        return ExitCode.COLLISIONS
-    elif orphans:
-        logger.warning("{} orphans found".format(len(orphans)))
-        print(dict(orphans_by_host))
-        return ExitCode.ORPHANS
-    else:
-        logger.info(
-            "No orphans found out of {} service registrations seen".format(
-                len(zk_instance_data)
+    exit_log = []
+
+    if check_collisions:
+        # collisions
+        instance_by_addr: DefaultDict[Tuple[str, int], Set[str]] = defaultdict(set)
+        for nerve_inst in nerve_instance_data:
+            instance_by_addr[(nerve_inst.host, nerve_inst.port)].add(nerve_inst.service)
+        collisions: List[str] = []
+        for zk_inst in zk_instance_data:
+            nerve_services = instance_by_addr[(zk_inst.host, zk_inst.port)]
+            if len(nerve_services) >= 1 and zk_inst.service not in nerve_services:
+                collisions.append(
+                    f"[{zk_inst.host}:{zk_inst.port}] {zk_inst.service} collides with {nerve_services}"
+                )
+
+        if collisions:
+            logger.warning("Collisions found! Traffic is being misrouted!")
+            print("\n".join(collisions))
+            return ExitCode.COLLISIONS
+        else:
+            exit_log.append(
+                f"No collisions found out of {len(zk_instance_data)} service registrations seen."
             )
-        )
-        return ExitCode.OK
+    if check_orphans:
+        orphans = zk_instance_data - nerve_instance_data
+
+        # groupby host
+        orphans_by_host: DefaultDict[str, List[Tuple[int, str]]] = defaultdict(list)
+        for orphan in orphans:
+            orphans_by_host[orphan.host].append((orphan.port, orphan.service))
+
+        if orphans:
+            logger.warning("{} orphans found".format(len(orphans)))
+            print(dict(orphans_by_host))
+            return ExitCode.ORPHANS
+        else:
+            exit_log.append(
+                f"No orphans found out of {len(zk_instance_data)} service registrations seen."
+            )
+
+    logger.info("\n".join(exit_log))
```

Comment:
just curious: why are we logging these at the same time rather than just as we hit those log statements?

Reply:
oh, that's a good call, ty! I'd organized this differently previously to only log+exit once but reverted everything else.